batch/sql/add-user-resources-ready-cores-token.sql

```diff
@@ -0,0 +1,176 @@
+SET @n_tokens = 200;
+
+ALTER TABLE globals ADD COLUMN n_tokens INT NOT NULL;
+UPDATE globals SET n_tokens = @n_tokens;
+
+CREATE TABLE IF NOT EXISTS `batches_staging` (
+  `batch_id` BIGINT NOT NULL,
+  `token` INT NOT NULL,
+  `n_jobs` INT NOT NULL DEFAULT 0,
+  `n_ready_jobs` INT NOT NULL DEFAULT 0,
+  `ready_cores_mcpu` BIGINT NOT NULL DEFAULT 0,
+  PRIMARY KEY (`batch_id`, `token`),
+  FOREIGN KEY (`batch_id`) REFERENCES batches(`id`) ON DELETE CASCADE
+) ENGINE = InnoDB;
+
+ALTER TABLE user_resources ADD COLUMN token INT NOT NULL DEFAULT 0;
+ALTER TABLE user_resources DROP PRIMARY KEY, ADD PRIMARY KEY(`user`, `token`);
+
+ALTER TABLE ready_cores ADD COLUMN token INT NOT NULL DEFAULT 0;
+ALTER TABLE ready_cores ADD PRIMARY KEY(`token`);
+
+CREATE TEMPORARY TABLE tmp_resources AS (
+  SELECT batch_id, user, closed,
+    0 as token,
+    COUNT(*) AS n_jobs,
+    COALESCE(SUM(state = 'Ready'), 0) AS n_ready_jobs,
+    COALESCE(SUM(state = 'Running'), 0) AS n_running_jobs,
+    COALESCE(SUM(IF(state = 'Ready', cores_mcpu, 0)), 0) AS ready_cores_mcpu,
+    COALESCE(SUM(IF(state = 'Running', cores_mcpu, 0)), 0) AS running_cores_mcpu
+  FROM jobs
+  INNER JOIN batches ON batches.id = jobs.batch_id
+  GROUP BY batch_id, user, closed
+);
+
+SELECT COALESCE(SUM(ready_cores_mcpu), 0) INTO @closed_ready_cores_mcpu
+FROM tmp_resources
+WHERE closed;
+
+UPDATE ready_cores SET ready_cores_mcpu = @closed_ready_cores_mcpu;
+
+UPDATE user_resources
+RIGHT JOIN (
+  SELECT user, token,
+    COALESCE(SUM(n_ready_jobs), 0) as n_ready_jobs,
+    COALESCE(SUM(n_running_jobs), 0) as n_running_jobs,
```

Comment:
Do we need casts here?

Reply:
I don't think so.  I think MySQL will implicitly convert between decimal an integer types.